src/main/java/software/amazon/awssdk/crt/http/HttpConnection.java

```diff
@@ -0,0 +1,189 @@
+
+        int port = uri.getPort();
+
+        /* Pick a default port based on the scheme if one wasn't set in the URI */
+        if (port == -1) {
+            if (HTTP.equals(uri.getScheme()))  { port = DEFAULT_HTTP_PORT; }
+            if (HTTPS.equals(uri.getScheme())) { port = DEFAULT_HTTPS_PORT; }
+        }
+
+        if (port <= 0 || 65535 < port) {
+            throw new IllegalArgumentException("Invalid or missing port: " + uri);
+        }
+
+        this.uri = uri;
+        this.port = port;
+        this.useTls = HTTPS.equals(uri.getScheme());
+        this.clientBootstrap = bootstrap;
+        this.socketOptions = socketOptions;
+        this.tlsContext = tlsContext;
+        this.connectedFuture = new CompletableFuture<>();
+        this.shutdownFuture = new CompletableFuture<>();
+    }
+
+    /**
+     * Schedules a connection task on the EventLoop to connect to the Http Endpoint
+     * @return Future indicating when the connection has completed.
+     */
+    private CompletableFuture<HttpConnection> connect() throws CrtRuntimeException {
+        if (!isNull()) {
+            return connectedFuture;
+        }
+
+        acquire(httpConnectionNew(this,
+                clientBootstrap.native_ptr(),
+                socketOptions.native_ptr(),
+                useTls ? tlsContext.native_ptr() : 0,
+                uri.getHost(),
+                port));
+
+        return connectedFuture;
+    }
+
+    /**
+     * Closes and frees this HttpConnection and any native sub-resources associated with this connection
+     */
+    @Override
+    public void close() {
+        if (!isNull()) {
+            httpConnectionRelease(release());
+        }
+
+        super.close();
+    }
+
+    /** Called from Native EventLoop when the connection is established the first time **/
+    private void onConnectionComplete(int errorCode) {
+        if (errorCode == AWS_CRT_SUCCESS) {
+            connectedFuture.complete(this);
+        } else {
+            /* The user can't close this HttpConnection object since they never got a reference to it through the
+                CompletableFuture, so we need to close ourselves on Connection Error. */
+            this.close();
+            connectedFuture.completeExceptionally(new HttpException(errorCode));
+        }
+    }
+
+    /** Called from Native EventLoop when the connection is shutdown. **/
+    private void onConnectionShutdown(int errorCode) {
+        if (errorCode == AWS_CRT_SUCCESS) {
+           shutdownFuture.complete(null);
+        } else {
+            shutdownFuture.completeExceptionally(new HttpException(errorCode));
+        }
+    }
+
+    /**
+     * Schedules a task on the Native EventLoop to shut down the current connection
+     * @return When this future completes, the shutdown is complete
+     */
+    public CompletableFuture<Void> shutdown() {
+        if (isNull()) {
+           return shutdownFuture;
```

Comment:
Since the `shutdownFuture` is private, users can't be notified of the event unless they trigger a shutdown themselves. Users might want to be notified ASAP of shutdowns that happen unexpectedly

Reply:
Added `getShutdownFuture()` method